### PR TITLE
otel: simple OTEL collector/Jaeger/Aspire stack for testing purposes

### PR DESCRIPTION
**Pull request by krissetto**
**- What I did**
Added a compose stack to be used for testing out OTEL functionality

**- How I did it**
- Added compose stack;
- Added basic OTEL collector config;
- Added basic usage doc

**- How to verify it**
Read the doc and try it out :)

**- Description for the changelog**

```markdown changelog
Add simple compose file for testing OTEL traces locally
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: hack/otel/compose.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  jaeger:
+    image: jaegertracing/all-in-one:latest
+    ports:
+      - 16686:16686
+
+  aspire-dashboard:
+    image: mcr.microsoft.com/dotnet/nightly/aspire-dashboard
+    ports:
+      - 18888:18888
```
Comment (thaJeztah) [inline hack/otel/compose.yaml:13]: wondering if we should use `127.0.0.1` for these; just in case someone runs this without considering that mapping the port will default to the public interface. @akerouanton any thoughts?

File: hack/otel/usage.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+- a Jaeger container to visualize traces;
+- an alternative Aspire Dashboard container to visualize traces;
+
+The OTEL collector is configured to export Traces to both the Jaeger and Aspire containers.
+
+The `hack/otel` directory contains the compose file with the services configured, along with a basic configuration file for the OTEL collector.
+
+## How can I use it?
+
+1) Export the env var used to override the OTLP endpoint:
```
Comment (thaJeztah) [inline hack/otel/usage.md:14]: Nit; should this be a actual markdown numbered list?

Comment (krissetto) [reply hack/otel/usage.md:14]: Mmh good question, I thought it looked a bit better but I guess it depends on the markdown renderer used. No hard feelings here, if it's "more correct" to do otherwise i'll change this

**Review by akerouanton — APPROVED** [review]

File: hack/otel/compose.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+name: moby-otel
+
+services:
+
+  jaeger:
+    image: jaegertracing/all-in-one:latest
```
Comment (akerouanton) [inline hack/otel/compose.yaml:6]: I think it's worth adding `restart: always` to every service.

Comment (krissetto) [reply hack/otel/compose.yaml:6]: Good idea. Reading the compose spec i see `always: The policy always restarts the container until its removal.`. 
Maybe using `unless-stopped` might be better in this case? I wouldn't want potential users (even if we kinda only expect devs and power-users to use this) to accidentally have these services running forever

Comment (akerouanton) [reply hack/otel/compose.yaml:6]: I thought `unless-stopped` wouldn't be right here because AFAIK if the container is stopped during dockerd shutdown, it won't automatically restart (eg. on a reboot). But I tried it and it seems to work properly, so we can also consider using that restart policy.

File: hack/otel/otelcol.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# Receive signals over gRPC and HTTP
+# moby currently uses http
+receivers:
+  otlp:
+    protocols:
+      grpc:
+      http:
```
Comment (akerouanton) [inline hack/otel/otelcol.yaml:7]: Just pulled `otel/opentelemetry-collector:latest` and it now defaults to `localhost`, so we need this instead:

```suggestion
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
```

File: hack/otel/usage.md (hunk 2 above)
Comment (krissetto) [reply hack/otel/usage.md:14]: sounds good to me! I'll fix up this pr today since it's been lingering for a while

Comment (thaJeztah) [reply hack/otel/usage.md:14]: I very much forgot it was still there 😂 ❤️

File: hack/otel/compose.yaml (hunk 1 above)
Comment (thaJeztah) [reply hack/otel/compose.yaml:13]: If we're updating; do you think it's better to be explicit (`0.0.0.0:18888:18888`) for these? Might give more of a clue that it's listening to "any" IP (not sure if we changed semantics on IPv4/IPv6 for that, but I think we still consider `0.0.0.0` to be "either IPv4/6")

**Comment by krissetto** [comment]
I addressed the feedback given here, @akerouanton if you can confirm it still works for your workflow after the port mapping and otelcol changes that'd be great

File: hack/otel/compose.yaml (hunk 1 above)
Comment (krissetto) [reply hack/otel/compose.yaml]: made this more explicit

File: hack/otel/compose.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+  aspire-dashboard:
+    image: mcr.microsoft.com/dotnet/nightly/aspire-dashboard
+    ports:
+      - 0.0.0.0:18888:18888
+    environment:
+      DOTNET_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS: 'true'
+  
+  otelcol:
+    image: otel/opentelemetry-collector:latest
```
Comment (akerouanton) [inline hack/otel/compose.yaml:18]: The `otel/opentelemetry-collector` image only contains a minimal set of processors, connectors, and no filters. So it's not particularly useful beyond just forwarding to multiple exporters.

Could you switch to `otel/opentelemetry-collector-contrib` instead? This would allow us to easily experiment with processing pipelines directly from this repo.

```suggestion
    image: otel/opentelemetry-collector-contrib:latest
```

File: hack/otel/compose.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    restart: always
+    depends_on:
+      - jaeger
+      - aspire-dashboard
+    ports:
+      - 4318:4318 # default otlp http port
+    volumes:
+      # Mount the otelcol.yml config file
+      - ./otelcol.yaml:/etc/otelcol/config.yaml
```
Comment (akerouanton) [inline hack/otel/compose.yaml:27]: For the same reason (ease experimentation from this repo), could you switch to a Compose watch?

```suggestion
    develop:
      watch:
        - action: sync+restart
          path: ./otelcol.yaml
          target: /etc/otelcol-contrib/config.yaml
```

File: hack/otel/compose.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+name: moby-otel
+
+services:
+
+  jaeger:
+    image: jaegertracing/all-in-one:latest
+    ports:
+      - 16686:16686
+
+  aspire-dashboard:
```
Comment (akerouanton) [inline hack/otel/compose.yaml:10]: Could you add `restart: always` to this service?

File: hack/otel/compose.yaml (hunk 3 above)
Comment (akerouanton) [reply hack/otel/compose.yaml:6]: Could you address this please?

File: hack/otel/compose.yaml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      - 0.0.0.0:18888:18888
+    environment:
+      DOTNET_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS: 'true'
+  
+  otelcol:
+    image: otel/opentelemetry-collector:latest
+    restart: always
+    depends_on:
+      - jaeger
+      - aspire-dashboard
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply hack/otel/compose.yaml:22]: AFAIK you can't manually import traces into Aspire, but you can with Jaeger. This is useful for instance when you want to load the OTel traces we dump at the end of integration tests on the CI. (I used that trick a couple of times to debug DNS-related issues on the CI)

OTOH, Aspire supports OTel logs and metrics. The Engine doesn't emit them, but this is useful if you want to experiment with a processing pipeline that transforms traces into metrics.

I'm planning to add a pipeline or two to have an easy way to track the performance of various operations in libnetwork (e.g. `libnetwork.drivers.bridge.createNetwork` and `daemon.connectToNetwork`) to better understand the impact of some of the changes we're making (e.g. `iptables` vs `nftables`).

Jaeger is also more "barebone", so I tend to use it for looking at traces, and then switching to Aspire to double-check that my pipelines are correctly exporting metrics. I could do everything with Aspire, so it's mostly personal preference.

Long story short, it's definitely useful to have both, and I don't think it'd add much value to have a Compose profile to choose which one to run.

[the remark this answers is not in the document]
Comment (krissetto) [reply hack/otel/compose.yaml:22]: i can include the profiles if we want that, and will implement the suggestions above as well!

[the remark this answers is not in the document]
Comment (thaJeztah) [reply hack/otel/compose.yaml:22]: Thanks! Perhaps in that case we don't need the profiles. I was mostly curious if we needed both, or if we could pick "the best one". Looks like there's still some parts not provided by both, so probably fine to keep'em both for now.

[the remark this answers is not in the document]
Comment (thaJeztah) [reply hack/otel/compose.yaml:22]: Slightly wondering if `contrib/` would be a more appropriate place for this compose-file 🤔

[the remark this answers is not in the document]
Comment (tianon) [reply hack/otel/compose.yaml:22]: > Slightly wondering if `contrib/` would be a more appropriate place for this compose-file 🤔

100% agree (hack is traditionally for scripts related to building, this is more of an example, which is what contrib is for)

**Comment by krissetto** [comment]
@thaJeztah could you add the `impact/` label so the CI isn't angry? or is it best to remove the changelog msg? it's a bit confusing

**Comment by thaJeztah** [comment]
Added the label; we can remove it later if we don't think it's worth including in the release notes 👍

File: contrib/otel/usage.md
Comment (thaJeztah) [inline contrib/otel/usage.md:1]: We can probably rename this to `README.md` then it shows up on Github if someone browses to the directory 😄

Comment (krissetto) [reply contrib/otel/usage.md:1]: done

**Comment by akerouanton** [comment]
I think we don't need a changelog entry for this one as it's not a user-facing change.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by akerouanton — APPROVED** [review]
